### ISSUE TITLE
stuck at some point during mikado pick

### ISSUE THREAD

**Issue by lijing28101** [EI-CoreBioinformatics/mikado#295]
Hi, I'm running mikado2 with padding on maize genome. And I tried same command with dame files in two hpc clusters. I totally have 685 chromosomes and scaffolds, 648 already finished according to the log. But both of them stuck at chromosome 6 for a long time.

1.Running with 32 cores and 360G memory. It has been run for almost 3 days, but it's stuck at some point for over 35 hours. Here are some last lines from log.
```
2020-03-07 23:02:19,269 - chr9:161591313-161850683 - locus.py:1339 - INFO - expand_transcript - LociProcesser-11 - SRR5186509.gtf_STRG.34374.2 has now start 161763925, end 161778780; CDS moved to 161778780, end 161765076.
2020-03-07 23:02:19,284 - chr9:161591313-161850683 - locus.py:215 - INFO - finalize_alternative_splicing - LociProcesser-11 - SRR1621016.gtf_STRG.28735.1 is now padded
2020-03-07 23:02:19,284 - chr9:161591313-161850683 - locus.py:215 - INFO - finalize_alternative_splicing - LociProcesser-11 - ERR965599.gtf_STRG.28300.2 is now padded
2020-03-07 23:02:19,284 - chr9:161591313-161850683 - locus.py:215 - INFO - finalize_alternative_splicing - LociProcesser-11 - SRR5186509.gtf_STRG.34374.2 is now padded
2020-03-07 23:02:19,284 - chr9:161591313-161850683 - locus.py:215 - INFO - finalize_alternative_splicing - LociProcesser-11 - SRR5166702.gtf_STRG.32727.1 is now padded
2020-03-07 23:02:21,686 - listener - picker.py:782 - INFO - __submit_multi_threading - MainProcess - Finished with chromosome chr9
2020-03-07 23:44:10,227 - chr6:17083781-17871432 - locus.py:1339 - INFO - expand_transcript - LociProcesser-25 - SRR6061626.gtf_STRG.21794.1 has now start 17309624, end 17311790.
2020-03-07 23:44:10,229 - chr6:17083781-17871432 - locus.py:215 - INFO - finalize_alternative_splicing - LociProcesser-25 - SRR6061626.gtf_STRG.21794.1 is now padded
2020-03-07 23:44:22,197 - chr6:17083781-17871432 - locus.py:1339 - INFO - expand_transcript - LociProcesser-25 - SRR1633470.gtf_STRG.15079.2 has now start 17249803, end 17261765.
2020-03-07 23:44:22,198 - chr6:17083781-17871432 - locus.py:215 - INFO - finalize_alternative_splicing - LociProcesser-25 - SRR1633470.gtf_STRG.15079.2 is now padded
```

2. Running with 120 cores and 1500G memory. This one is running almost 2 days, and also stuck at chromosome 6 for 13 hours. But this cluster is stuck at a transcript previous to the first one.
```
2020-03-08 14:03:35,828 - chr9:161591313-161850683 - locus.py:1339 - INFO - expand_transcript - LociProcesser-1 - SRR5368988.gtf_STRG.27419.3 has now start 161763925, end 161778784; CDS moved to 161778784, end 161764235.
2020-03-08 14:03:36,176 - chr9:161591313-161850683 - locus.py:1339 - INFO - expand_transcript - LociProcesser-1 - ERR965599.gtf_STRG.28300.2 has now start 161763925, end 161778780; CDS moved to 161778780, end 161764235.
2020-03-08 14:03:36,253 - chr9:161591313-161850683 - locus.py:1339 - INFO - expand_transcript - LociProcesser-1 - SRR5186509.gtf_STRG.34374.2 has now start 161763925, end 161778780; CDS moved to 161778780, end 161765076.
2020-03-08 14:03:36,268 - chr9:161591313-161850683 - locus.py:215 - INFO - finalize_alternative_splicing - LociProcesser-1 - SRR1621016.gtf_STRG.28735.1 is now padded
2020-03-08 14:03:36,268 - chr9:161591313-161850683 - locus.py:215 - INFO - finalize_alternative_splicing - LociProcesser-1 - ERR965599.gtf_STRG.28300.2 is now padded
2020-03-08 14:03:36,268 - chr9:161591313-161850683 - locus.py:215 - INFO - finalize_alternative_splicing - LociProcesser-1 - SRR5186509.gtf_STRG.34374.2 is now padded
2020-03-08 14:03:36,268 - chr9:161591313-161850683 - locus.py:215 - INFO - finalize_alternative_splicing - LociProcesser-1 - SRR5166702.gtf_STRG.32727.1 is now padded
2020-03-08 14:03:43,975 - listener - picker.py:782 - INFO - __submit_multi_threading - MainProcess - Finished with chromosome chr9
2020-03-08 21:04:21,163 - chr6:17083781-17871432 - locus.py:1339 - INFO - expand_transcript - LociProcesser-15 - SRR6061626.gtf_STRG.21794.1 has now start 17309624, end 17311790.
2020-03-08 21:04:21,228 - chr6:17083781-17871432 - locus.py:215 - INFO - finalize_alternative_splicing - LociProcesser-15 - SRR6061626.gtf_STRG.21794.1 is now padded
```
Do you have any idea about the problem?

Thanks,
Jing

**Comment by lucventurini** [comment]
Hi @lijing28101 

Why the extremely high memory values? Is it going out of memory otherwise?

It's very hard for me to understand what is going badly only from this, unfortunately. I would need to know the version of Mikado that you are using.

If you are using 1.2.4, the likeliest explanation is that Mikado got stuck in a dense locus with a lot of transcripts. The original version of Mikado unfortunately relied on quadratic functions that scale really poorly for dense loci. We should have solved that in the development versions.

**Comment by lijing28101** [comment]
Hi @lucventurini,
My data is large, the mikado.db from serialize is about 7G. I'm not sure how many memory the mikado pick will use, so I set a high memory.
I used mikado v2.0rc6, not the old one.

**Comment by lucventurini** [comment]
Hi @lijing28101 

Most probably there are some loci that are extra dense. I'm a little bit disappointed as I thought I had removed any bottleneck of this kind. The database is not too big to be honest, we've had files of that size. 

My best suggestion is to use `gffread` on `mikado_prepared.gtf` to remove redundant intron chains. It's not ideal (hypothetically the best model could be discarded in some loci with this crude procedure) but it should drastically reduce the number of transcripts to evaluate.

Specifically the options to use should be `gffread -T -M -Q -d mikado_prepared.duplicated.txt -o mikado_prepared.nr.gtf mikado_prepared.gtf`.

You do **not** need to regenerate the data for `mikado serialise`.

I would be interested in having the `mikado_prepared.duplicated.txt`file from above if possible.

**Comment by lijing28101** [comment]
Hi @lucventurini ,

Thanks for your suggestion. I run gffread and find many duplication. Because I merged stringtie output from over 1000 samples. The `mikado_prepare.gtf` reduce from 48G to 3.4G. The `mikado_prepared.duplicated.txt` is 2.9G, are you sure you want it?
And I tried `mikado_prepared.nr.gtf` for mikado pick directly without rerun mikado serialise. But there is error:
```
2020-03-09 20:16:24,977 - listener - picker.py:896 - CRITICAL - __parse_multithreaded - MainProcess - Transcript SRR3098753.gtf_STRG.5.1 is invalid
2020-03-09 20:16:25,030 - listener - picker.py:1150 - ERROR - __call__ - MainProcess - The input files were not properly sorted! Please run prepare and retry.
Exception ignored in: <finalize object at 0x2ac3f85898f0; dead>
Traceback (most recent call last):
  File "/work/LAS/mash-lab/jing/bin/anaconda3/envs/mikado2/lib/python3.7/weakref.py", line 552, in __call__
    return info.func(*info.args, **(info.kwargs or {}))
  File "/work/LAS/mash-lab/jing/bin/anaconda3/envs/mikado2/lib/python3.7/tempfile.py", line 795, in _cleanup
    _shutil.rmtree(name)
  File "/work/LAS/mash-lab/jing/bin/anaconda3/envs/mikado2/lib/python3.7/shutil.py", line 495, in rmtree
    onerror(os.rmdir, path, sys.exc_info())
  File "/work/LAS/mash-lab/jing/bin/anaconda3/envs/mikado2/lib/python3.7/shutil.py", line 493, in rmtree
    os.rmdir(path)
OSError: [Errno 39] Directory not empty: '/work/LAS/mash-lab/jing/maize/mikado_20/out_new/mikado_pick_tmpr1k10qv7'
```
Need I rerun blast, transdecoder and serialize?

**Comment by lucventurini** [comment]
Hi @lijing28101 

No, the problem is that the GTF is not sorted properly after the `gffread` filtering. I have to think on how to perform the sorting, I will let you know ASAP.

Thanks for your patience

**Comment by srividya22** [comment]
@lijing28101 

How long did serialize take on your data ?  Can you post your serialize log how many target queries are loaded 

2020-03-07 21:41:29,048 - serialiser - xml_serialiser.py:331 - INFO - __serialize_targets - MainProcess - Started to serialise the targets
2020-03-07 21:41:40,602 - serialiser - xml_serialiser.py:366 - INFO - __serialize_targets - MainProcess - Loaded 558898 objects into the "target" table
2020-03-07 21:41:40,610 - serialiser - xml_serialiser.py:260 - INFO - __serialize_queries - MainProcess - Started to serialise the queries
2020-03-07 21:41:40,883 - serialiser - xml_serialiser.py:308 - INFO - __serialize_queries - MainProcess - Loaded 0 objects into the "query" table

Mine says 558898 objects in target and 0 queries. Not sure if this is any problem. serialize finished quite fast. 

Also can you check you job stderr log for mikado pick and see any errors popping there ? 

mikado pick log for me looked very similar to yours.

LociProcesser-2 - CLASS2_SRR7066591_Super-Scaffold_61.1107.0 is now padded and has now start 13450061, end 13461035.
2020-03-07 21:54:15,680 - Super-Scaffold_61:12983644-13891052 - locus.py:250 - INFO - finalize_alternative_splicing - LociProcesser-2 - CLASS2_SRR7066589_Super-Scaffold_61.1086.0 is now padded and has now start 13450061, end 13461035.
2020-03-07 21:54:16,275 - Super-Scaffold_61:12983644-13891052 - locus.py:250 - INFO - finalize_alternative_splicing - LociProcesser-2 - stringtie_SRR7066590_STRG.6340.1 is now padded and has now start 13731994, end 13736199.
2020-03-07 21:54:16,275 - Super-Scaffold_61:12983644-13891052 - locus.py:250 - INFO - finalize_alternative_splicing - LociProcesser-2 - CLASS2_SRR7066589_Super-Scaffold_61.1104.0 is now padded and has now start 13731994, end 13736199.
2020-03-07 21:54:16,275 - Super-Scaffold_61:12983644-13891052 - locus.py:250 - INFO - finalize_alternative_splicing - LociProcesser-2 - stringtie_SRR7066591_STRG.6737.2 is now padded and has now start 13731994, end 13736199.
2020-03-07 21:54:16,961 - listener - picker.py:785 - INFO - __submit_multi_threading - MainProcess - Finished with chromosome Super-Scaffold_61

**Comment by lijing28101** [comment]
@srividya22 serialize took about 10h using the version before merging #280. And I didn't see any error for pick. It's still stuck.
```
2020-03-05 13:47:29,012 - serialiser - serialise.py:290 - INFO - setup - MainProcess - Command line: /work/LAS/mash-lab/jing/bin/anaconda3/envs/mikado2/bin/mikado serialise --start-method spawn --procs 64 --blast_targets Zm-B73-REFERENCE-NAM-5.0.fa --json-conf configuration_20.yaml --xml mikado.blast.xml --orfs transdecoder.bed
2020-03-05 13:47:29,012 - serialiser - serialise.py:296 - INFO - setup - MainProcess - Random seed: 0
2020-03-05 13:47:29,015 - serialiser - serialise.py:334 - INFO - setup - MainProcess - Using a sqlite database (location: /work/LAS/mash-lab/jing/maize/mikado_20/mikado.db)
2020-03-05 13:47:29,015 - serialiser - serialise.py:338 - INFO - setup - MainProcess - Requested 64 threads, forcing single thread: False
2020-03-05 13:47:29,015 - serialiser - serialise.py:140 - INFO - load_orfs - MainProcess - Starting to load ORF data
2020-03-05 21:52:12,092 - serialiser - orf.py:449 - INFO - __serialize_multiple_threads - MainProcess - Finished loading 43118643 ORFs into the database
2020-03-05 22:06:28,483 - serialiser - serialise.py:152 - INFO - load_orfs - MainProcess - Finished loading ORF data
2020-03-05 22:06:46,412 - serialiser - serialise.py:106 - INFO - load_blast - MainProcess - Starting to load BLAST data
2020-03-05 22:25:46,682 - serialiser - xml_serialiser.py:327 - INFO - __serialize_targets - MainProcess - Started to serialise the targets
2020-03-05 22:25:46,689 - serialiser - xml_serialiser.py:362 - INFO - __serialize_targets - MainProcess - Loaded 685 objects into the "target" table
2020-03-05 22:25:46,689 - serialiser - xml_serialiser.py:256 - INFO - __serialize_queries - MainProcess - Started to serialise the queries
2020-03-05 22:26:09,369 - serialiser - xml_serialiser.py:304 - INFO - __serialize_queries - MainProcess - Loaded 0 objects into the "query" table
2020-03-05 22:28:40,834 - serialiser - xml_serialiser.py:526 - INFO - serialize - MainProcess - Loaded 0 alignments for 0 queries
2020-03-05 22:28:40,834 - serialiser - xml_serialiser.py:528 - INFO - serialize - MainProcess - Finished loading blast hits
2020-03-05 22:28:49,778 - serialiser - serialise.py:125 - INFO - load_blast - MainProcess - Finished to load BLAST data
2020-03-05 22:28:49,778 - serialiser - serialise.py:81 - INFO - load_junctions - MainProcess - Starting to load junctions: ['merged20.junctions.bed']
2020-03-05 22:29:10,161 - serialiser - serialise.py:91 - INFO - load_junctions - MainProcess - Loaded junctions
2020-03-05 22:29:10,161 - serialiser - serialise.py:383 - INFO - serialise - MainProcess - Finished

```

**Comment by srividya22** [comment]
@lucventurini 

I have a question regarding how mikado handles transcripts from several RNASeq samples.  Does mikado expect RNASeq transcripts from stringtie needs to me merged  using  ( say stringtie merge or taco ) before running mikado. Would that help solve the issue of duplication ?

**Comment by lucventurini** [comment]
@srividya22 @lijing28101 

Mikado by itself actually does not expect any redundancy removal to have been performed beforehand. That said, even while we put a lot of effort into making it run in ~linear time for the size of the inputs, it was not originally written with the idea of handling such deep datasets.

In #270, @swarbred and I discussed about introducing intron-chain redundancy removal into `mikado prepare`. It was not a priority for us, but evidently this is something that people like you need and urgently.

I am currently directing a lot of my efforts to close #280, but I will try to dedicate some time this week to finding a way to help you finding a way to sort the files after using `gffread`.

**Comment by srividya22** [comment]
@lucventurini  Thanks a lot and really appreciate your help to resolve this issue.

**Comment by lijing28101** [comment]
Hi @lucventurini, after stucking 9 days, mikado pick finish all chromosomes, but have error message
```
2020-03-16 16:29:07,084 - listener - picker.py:786 - INFO - __submit_multi_threading - MainProcess - Joined children processes; starting to merge partial files
2020-03-16 16:49:04,860 - listener - loci_processer.py:94 - INFO - merge_loci - MainProcess - We have a total of 569307 genes
2020-03-16 16:49:11,732 - main - __init__.py:120 - ERROR - main - MainProcess - Mikado crashed, cause:
2020-03-16 16:49:11,732 - main - __init__.py:121 - ERROR - main - MainProcess - manage_index() got an unexpected keyword argument 'dumps'
multiprocessing.pool.RemoteTraceback:
"""
Traceback (most recent call last):
  File "/work/LAS/mash-lab/jing/bin/anaconda3/envs/mikado2/lib/python3.7/multiprocessing/pool.py", line 121, in worker
    result = (True, func(*args, **kwds))
  File "/work/LAS/mash-lab/jing/bin/anaconda3/envs/mikado2/lib/python3.7/multiprocessing/pool.py", line 44, in mapstar
    return list(map(*args))
TypeError: manage_index() got an unexpected keyword argument 'dumps'
"""

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/work/LAS/mash-lab/jing/bin/anaconda3/envs/mikado2/lib/python3.7/site-packages/Mikado/__init__.py", line 106, in main
    args.func(args)
  File "/work/LAS/mash-lab/jing/bin/anaconda3/envs/mikado2/lib/python3.7/site-packages/Mikado/subprograms/pick.py", line 222, in pick
    regions = None
  File "/work/LAS/mash-lab/jing/bin/anaconda3/envs/mikado2/lib/python3.7/site-packages/Mikado/picking/picker.py", line 1147, in __call__
    if self.json_conf["db_settings"]["dbtype"] == "sqlite":
  File "/work/LAS/mash-lab/jing/bin/anaconda3/envs/mikado2/lib/python3.7/site-packages/Mikado/picking/picker.py", line 1113, in _parse_and_submit_input
    """
  File "/work/LAS/mash-lab/jing/bin/anaconda3/envs/mikado2/lib/python3.7/site-packages/Mikado/picking/picker.py", line 793, in __submit_multi_threading
    handles,
  File "/work/LAS/mash-lab/jing/bin/anaconda3/envs/mikado2/lib/python3.7/site-packages/Mikado/picking/loci_processer.py", line 99, in merge_loci
    for row in sub_scores_rows:
  File "/work/LAS/mash-lab/jing/bin/anaconda3/envs/mikado2/lib/python3.7/multiprocessing/pool.py", line 325, in <genexpr>
    return (item for chunk in result for item in chunk)
  File "/work/LAS/mash-lab/jing/bin/anaconda3/envs/mikado2/lib/python3.7/multiprocessing/pool.py", line 748, in next
    raise value
TypeError: manage_index() got an unexpected keyword argument 'dumps'
Traceback (most recent call last):
  File "/work/LAS/mash-lab/jing/bin/anaconda3/envs/mikado2/lib/python3.7/weakref.py", line 628, in _exitfunc
    f()
  File "/work/LAS/mash-lab/jing/bin/anaconda3/envs/mikado2/lib/python3.7/weakref.py", line 552, in __call__
    return info.func(*info.args, **(info.kwargs or {}))
  File "/work/LAS/mash-lab/jing/bin/anaconda3/envs/mikado2/lib/python3.7/tempfile.py", line 795, in _cleanup
    _shutil.rmtree(name)
  File "/work/LAS/mash-lab/jing/bin/anaconda3/envs/mikado2/lib/python3.7/shutil.py", line 495, in rmtree
    onerror(os.rmdir, path, sys.exc_info())
  File "/work/LAS/mash-lab/jing/bin/anaconda3/envs/mikado2/lib/python3.7/shutil.py", line 493, in rmtree
    os.rmdir(path)
OSError: [Errno 39] Directory not empty: '/work/LAS/mash-lab/jing/maize/mikado_20/out/mikado_pick_tmpcftn36h_'
```

**Comment by lucventurini** [comment]
Hi @lijing28101 

We are working at the moment on #270 which should greatly reduce the amount of input assemblies for `mikado pick`.

In general though, I cannot diagnose the bug you are encountering here. I checked the code and in no place the function `manage_index` uses the `dumps=` keyword that triggers the error.

I fear therefore that you might have been hit by a bug that I have since solved, as your run took 9 days.

If I may, I would recommend extracting all the models of a single chromosome/scaffold, then running the *up-to-date* version of Mikado on that subset of the data. This should tell us quickly whether this bug or another one is still present.

**Comment by lijing28101** [comment]
Hi @lucventurini 
Do you merge the new version to master branch?
I need to run mikado from prepare, right?

**Comment by lijing28101** [comment]
Hi @lucventurini 
I tried chromosome 6, which is stuck in the old version.
It finished very fast, and reduce to less than 1/10 of mikado_prepare.fasta than the old version.
Thanks for solving my problem!

**Comment by lucventurini** [comment]
This should now be fixed in version 2.0rc2 on both [PyPI](https://pypi.org/project/Mikado/2.0rc2/) and [BioConda](https://anaconda.org/bioconda/mikado). Please repoen if the issue represents itself with the new version!
